File: src/task-queue.ts
/** Filesystem-based task queue. Tasks are JSON files moved between directories. */

import { join } from "@std/path";

export type TaskState = "pending" | "current" | "complete" | "rejected";
export type TaskCounts = Record<TaskState, number>;

export const TASK_STATES: TaskState[] = [
  "pending",
  "current",
  "complete",
  "rejected",
];

export interface TaskSeed {
  summary: string;
  description?: string;
  files?: string[];
  acceptance?: string[];
  depends_on?: string[];
}

export interface Task {
  id: string;
  summary: string;
  description: string;
  files?: string[];
  acceptance?: string[];
  depends_on?: string[];
  assignee?: string;
  feedback?: string;
  createdBy?: string;
  retries: number;
}

export interface TaskEntry {
  task: Task;
  state: TaskState;
}

const TASKS_DIR = ".jackops/tasks";
let _idCounter = 0;

/** Generate a unique task ID. */
export function generateId(): string {
  return `task-${Date.now()}-${_idCounter++}`;
}

function stateDir(base: string, state: TaskState): string {
  return join(base, TASKS_DIR, state);
}

function taskPath(base: string, state: TaskState, id: string): string {
  return join(base, TASKS_DIR, state, `${id}.json`);
}

/** Create the tasks/ directory structure. */
export async function init(base: string): Promise<void> {
  for (const state of TASK_STATES) {
    await Deno.mkdir(stateDir(base, state), { recursive: true });
  }
}

/** Write a new task to pending/. */
export async function create(
  base: string,
  task: Omit<Task, "retries">,
): Promise<Task> {
  const full: Task = { ...task, retries: 0 };
  const path = taskPath(base, "pending", full.id);
  await Deno.writeTextFile(path, JSON.stringify(full, null, 2) + "\n");
  return full;
}

/** Read a task file. */
async function readTask(path: string): Promise<Task> {
  const text = await Deno.readTextFile(path);
  return JSON.parse(text) as Task;
}

/** Atomically move a task between states and optionally update fields. */
async function moveTask(
  base: string,
  id: string,
  from: TaskState,
  to: TaskState,
  update?: Partial<Task> | ((task: Task) => Partial<Task>),
): Promise<Task> {
  const src = taskPath(base, from, id);
  const dst = taskPath(base, to, id);
  const task = await readTask(src);
  const patch = typeof update === "function" ? update(task) : update;
  const updated = patch ? { ...task, ...patch } : task;
  // Write to destination first, then remove source.
  // If we crash between write and remove, we have a duplicate --
  // but that's better than losing the task.
  await Deno.writeTextFile(dst, JSON.stringify(updated, null, 2) + "\n");
  await Deno.remove(src);
  return updated;
}

/** Claim a pending task for a worker. Moves pending/ -> current/. */
export function claim(
  base: string,
  taskId: string,
  assignee: string,
): Promise<Task> {
  return moveTask(base, taskId, "pending", "current", { assignee });
}

/** Mark a current task as complete. Moves current/ -> complete/. */
export function complete(base: string, taskId: string): Promise<Task> {
  return moveTask(base, taskId, "current", "complete");
}

/** Unclaim a current task, moving it back to pending. */
export function unclaim(base: string, taskId: string): Promise<Task> {
  return moveTask(base, taskId, "current", "pending", {
    assignee: undefined,
  });
}

/** Reject a current task with feedback. Moves current/ -> rejected/. */
export function reject(
  base: string,
  taskId: string,
  feedback: string,
): Promise<Task> {
  return moveTask(base, taskId, "current", "rejected", { feedback });
}

/** Retry a rejected task. Moves rejected/ -> pending/, increments retries. */
export function retry(base: string, taskId: string): Promise<Task> {
  return moveTask(base, taskId, "rejected", "pending", (task) => ({
    retries: task.retries + 1,
    assignee: undefined,
    feedback: undefined,
  }));
}

/** List tasks in a given state (or all states). */
export async function list(
  base: string,
  state?: TaskState,
): Promise<TaskEntry[]> {
  const states = state ? [state] : TASK_STATES;
  const entries: TaskEntry[] = [];

  for (const s of states) {
    const dir = stateDir(base, s);
    try {
      for await (const entry of Deno.readDir(dir)) {
        if (!entry.isFile || !entry.name.endsWith(".json")) continue;
        const task = await readTask(join(dir, entry.name));
        entries.push({ task, state: s });
      }
    } catch (e) {
      if (e instanceof Deno.errors.NotFound) continue;
      throw e;
    }
  }

  return entries;
}

/** Get a specific task by ID, searching all states. */
export async function get(
  base: string,
  taskId: string,
): Promise<TaskEntry | null> {
  for (const s of TASK_STATES) {
    try {
      const task = await readTask(taskPath(base, s, taskId));
      return { task, state: s };
    } catch (e) {
      if (e instanceof Deno.errors.NotFound) continue;
      throw e;
    }
  }
  return null;
}

/** List pending tasks whose dependencies are all complete. */
export async function ready(base: string): Promise<Task[]> {
  const [pending, completed] = await Promise.all([
    list(base, "pending"),
    list(base, "complete"),
  ]);
  const completeIds = new Set(completed.map((e) => e.task.id));

  return pending
    .filter((e) => {
      const deps = e.task.depends_on ?? [];
      return deps.every((d) => completeIds.has(d));
    })
    .map((e) => e.task);
}

/** Count tasks by state (counts dir entries, avoids parsing JSON). */
export async function counts(
  base: string,
): Promise<TaskCounts> {
  const result: TaskCounts = {
    pending: 0,
    current: 0,
    complete: 0,
    rejected: 0,
  };
  for (const s of TASK_STATES) {
    const dir = stateDir(base, s);
    try {
      for await (const entry of Deno.readDir(dir)) {
        if (entry.isFile && entry.name.endsWith(".json")) {
          result[s]++;
        }
      }
    } catch (e) {
      if (e instanceof Deno.errors.NotFound) continue;
      throw e;
    }
  }
  return result;
}

/** Seed tasks from config, skipping duplicates by summary. Returns count seeded. */
export async function seed(base: string, tasks: TaskSeed[]): Promise<number> {
  await init(base);
  const existing = await list(base);
  const existingSummaries = new Set(existing.map((e) => e.task.summary));
  let seeded = 0;
  for (const t of tasks) {
    if (existingSummaries.has(t.summary)) continue;
    await create(base, {
      id: generateId(),
      summary: t.summary,
      description: t.description ?? t.summary,
      files: t.files,
      acceptance: t.acceptance,
      depends_on: t.depends_on,
    });
    seeded++;
  }
  return seeded;
}

File: src/taskqueue/taskqueue.go
package taskqueue

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"sync/atomic"
	"time"

	"github.com/AdriaanVE/jack-in/src/domain"
)

// safeTaskID matches valid task IDs: alphanumeric, hyphens, underscores, dots.
// Prevents path traversal via crafted task IDs passed through CLI args.
var safeTaskID = regexp.MustCompile(`^[a-zA-Z0-9_.-]+$`)

// ValidateTaskID checks that a task ID is safe for use in file paths.
func ValidateTaskID(id string) error {
	if id == "" {
		return fmt.Errorf("empty task ID")
	}
	if !safeTaskID.MatchString(id) {
		return fmt.Errorf("invalid task ID %q: must be alphanumeric, hyphens, underscores, dots", id)
	}
	return nil
}

const tasksDir = ".jack-in/tasks"

var idCounter atomic.Int64

// Task represents a unit of work in the queue.
type Task struct {
	ID             string    `json:"id"`
	Summary        string    `json:"summary"`
	Description    string    `json:"description"`
	Files          []string  `json:"files,omitempty"`
	Acceptance     []string  `json:"acceptance,omitempty"`
	DependsOn      []string  `json:"depends_on,omitempty"`
	Assignee       string    `json:"assignee,omitempty"`
	Feedback       string    `json:"feedback,omitempty"`
	CreatedBy      string    `json:"created_by,omitempty"`
	Retries        int       `json:"retries"`
	TransitionedAt time.Time `json:"transitioned_at,omitempty"`
}

// TaskEntry pairs a task with its current state.
type TaskEntry struct {
	Task  Task
	State domain.TaskState
}

// GenerateID creates a unique task ID.
func GenerateID() string {
	return fmt.Sprintf("task-%d-%d", time.Now().UnixMilli(), idCounter.Add(1)-1)
}

func stateDir(base string, state domain.TaskState) string {
	return filepath.Join(base, tasksDir, string(state))
}

func taskPath(base string, state domain.TaskState, id string) string {
	return filepath.Join(base, tasksDir, string(state), id+".json")
}

// TaskFilePath returns the file path for a task in a given state (exported for CLI use).
func TaskFilePath(base string, state domain.TaskState, id string) string {
	return taskPath(base, state, id)
}

// Init creates the task queue directory structure.
func Init(base string) error {
	for _, s := range domain.TaskStates {
		if err := os.MkdirAll(stateDir(base, s), 0o755); err != nil {
			return fmt.Errorf("creating task dir %s: %w", s, err)
		}
	}
	return nil
}

// Add creates a new task in the pending state.
func Add(base string, summary, description string) (*Task, error) {
	if description == "" {
		description = summary
	}
	task := Task{
		ID:             GenerateID(),
		Summary:        summary,
		Description:    description,
		TransitionedAt: time.Now(),
	}
	if err := writeTask(base, domain.StatePending, &task); err != nil {
		return nil, err
	}
	return &task, nil
}

// Get finds a task by ID, searching all states.
func Get(base string, taskID string) (*TaskEntry, error) {
	if err := ValidateTaskID(taskID); err != nil {
		return nil, err
	}
	for _, s := range domain.TaskStates {
		task, err := readTask(taskPath(base, s, taskID))
		if err != nil {
			if os.IsNotExist(err) {
				continue
			}
			return nil, err
		}
		return &TaskEntry{Task: *task, State: s}, nil
	}
	return nil, nil
}

// List returns tasks in a given state.
func List(base string, state domain.TaskState) ([]TaskEntry, error) {
	return listStates(base, []domain.TaskState{state})
}

// ListAll returns tasks across all states.
func ListAll(base string) ([]TaskEntry, error) {
	return listStates(base, domain.TaskStates)
}

func listStates(base string, states []domain.TaskState) ([]TaskEntry, error) {
	var entries []TaskEntry
	for _, s := range states {
		dir := stateDir(base, s)
		dirEntries, err := os.ReadDir(dir)
		if err != nil {
			if os.IsNotExist(err) {
				continue
			}
			return nil, err
		}
		for _, de := range dirEntries {
			if de.IsDir() || !strings.HasSuffix(de.Name(), ".json") {
				continue
			}
			task, err := readTask(filepath.Join(dir, de.Name()))
			if err != nil {
				return nil, err
			}
			entries = append(entries, TaskEntry{Task: *task, State: s})
		}
	}
	return entries, nil
}

// Counts returns the number of tasks in each state without parsing JSON.
func Counts(base string) (domain.TaskCounts, error) {
	var c domain.TaskCounts
	for _, s := range domain.TaskStates {
		dir := stateDir(base, s)
		dirEntries, err := os.ReadDir(dir)
		if err != nil {
			if os.IsNotExist(err) {
				continue
			}
			return c, err
		}
		n := 0
		for _, de := range dirEntries {
			if !de.IsDir() && strings.HasSuffix(de.Name(), ".json") {
				n++
			}
		}
		switch s {
		case domain.StatePending:
			c.Pending = n
		case domain.StateCurrent:
			c.Current = n
		case domain.StateReview:
			c.Review = n
		case domain.StateComplete:
			c.Complete = n
		case domain.StateRejected:
			c.Rejected = n
		}
	}
	return c, nil
}

// Validate checks for duplicate task files across states and returns a list of issues.
// Duplicates can occur if a transition write succeeds but the source removal fails.
func Validate(base string) ([]string, error) {
	var issues []string
	taskStates := make(map[string][]domain.TaskState)

	for _, s := range domain.TaskStates {
		dir := stateDir(base, s)
		dirEntries, err := os.ReadDir(dir)
		if err != nil {
			if os.IsNotExist(err) {
				continue
			}
			return nil, err
		}
		for _, de := range dirEntries {
			if de.IsDir() || !strings.HasSuffix(de.Name(), ".json") {
				continue
			}
			taskID := strings.TrimSuffix(de.Name(), ".json")
			taskStates[taskID] = append(taskStates[taskID], s)
		}
	}

	for taskID, states := range taskStates {
		if len(states) > 1 {
			stateNames := make([]string, len(states))
			for i, s := range states {
				stateNames[i] = string(s)
			}
			issues = append(issues, fmt.Sprintf("task %s exists in multiple states: %v", taskID, stateNames))
		}
	}
	return issues, nil
}

// FixDuplicates removes stale duplicate task files, keeping the most recent state.
// Returns the number of duplicates fixed.
func FixDuplicates(base string) (int, error) {
	taskStates := make(map[string][]domain.TaskState)

	for _, s := range domain.TaskStates {
		dir := stateDir(base, s)
		dirEntries, err := os.ReadDir(dir)
		if err != nil {
			if os.IsNotExist(err) {
				continue
			}
			return 0, err
		}
		for _, de := range dirEntries {
			if de.IsDir() || !strings.HasSuffix(de.Name(), ".json") {
				continue
			}
			taskID := strings.TrimSuffix(de.Name(), ".json")
			taskStates[taskID] = append(taskStates[taskID], s)
		}
	}

	fixed := 0
	// State priority: later in the lifecycle wins (rejected > complete > review > current > pending)
	statePriority := map[domain.TaskState]int{
		domain.StatePending:  0,
		domain.StateCurrent:  1,
		domain.StateReview:   2,
		domain.StateComplete: 3,
		domain.StateRejected: 4,
	}

	for taskID, states := range taskStates {
		if len(states) <= 1 {
			continue
		}
		// Find highest priority state (most recent in lifecycle)
		var keepState domain.TaskState
		maxPriority := -1
		for _, s := range states {
			if p := statePriority[s]; p > maxPriority {
				maxPriority = p
				keepState = s
			}
		}
		// Remove all other copies
		for _, s := range states {
			if s != keepState {
				path := taskPath(base, s, taskID)
				if err := os.Remove(path); err != nil {
					return fixed, fmt.Errorf("removing duplicate %s from %s: %w", taskID, s, err)
				}
				fixed++
			}
		}
	}
	return fixed, nil
}

// Ready returns pending tasks whose dependencies are all complete.
func Ready(base string) ([]Task, error) {
	pending, err := List(base, domain.StatePending)
	if err != nil {
		return nil, err
	}
	completed, err := List(base, domain.StateComplete)
	if err != nil {
		return nil, err
	}

	completeIDs := make(map[string]bool, len(completed))
	completeSummaries := make(map[string]bool, len(completed))
	for _, e := range completed {
		completeIDs[e.Task.ID] = true
		completeSummaries[e.Task.Summary] = true
	}

	var ready []Task
	for _, e := range pending {
		allDeps := true
		for _, dep := range e.Task.DependsOn {
			if !completeIDs[dep] && !completeSummaries[dep] {
				allDeps = false
				break
			}
		}
		if allDeps {
			ready = append(ready, e.Task)
		}
	}
	return ready, nil
}

// TaskSeed is the subset of fields used when seeding tasks from config.
type TaskSeed struct {
	Summary     string
	Description string
	Files       []string
	Acceptance  []string
	DependsOn   []string
}

// Seed creates tasks from config seeds, skipping duplicates by summary.
// Returns the number of tasks seeded.
func Seed(base string, seeds []TaskSeed) (int, error) {
	if err := Init(base); err != nil {
		return 0, err
	}

	existing, err := ListAll(base)
	if err != nil {
		return 0, err
	}

	existingSummaries := make(map[string]bool, len(existing))
	summaryToID := make(map[string]string, len(existing))
	for _, e := range existing {
		existingSummaries[e.Task.Summary] = true
		summaryToID[e.Task.Summary] = e.Task.ID
	}

	seeded := 0
	for _, s := range seeds {
		if existingSummaries[s.Summary] {
			continue
		}

		id := GenerateID()
		desc := s.Description
		if desc == "" {
			desc = s.Summary
		}

		// Resolve depends_on summary strings to task IDs
		var resolvedDeps []string
		for _, dep := range s.DependsOn {
			if resolved, ok := summaryToID[dep]; ok {
				resolvedDeps = append(resolvedDeps, resolved)
			} else {
				resolvedDeps = append(resolvedDeps, dep)
			}
		}

		task := Task{
			ID:          id,
			Summary:     s.Summary,
			Description: desc,
			Files:       s.Files,
			Acceptance:  s.Acceptance,
			DependsOn:   resolvedDeps,
		}
		if err := writeTask(base, domain.StatePending, &task); err != nil {
			return seeded, err
		}
		summaryToID[s.Summary] = id
		existingSummaries[s.Summary] = true
		seeded++
	}
	return seeded, nil
}

func readTask(path string) (*Task, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var t Task
	if err := json.Unmarshal(data, &t); err != nil {
		return nil, fmt.Errorf("parsing task %s: %w", path, err)
	}
	return &t, nil
}

func writeTask(base string, state domain.TaskState, task *Task) error {
	data, err := json.MarshalIndent(task, "", "  ")
	if err != nil {
		return err
	}
	data = append(data, '\n')
	return os.WriteFile(taskPath(base, state, task.ID), data, 0o644)
}
